### PR TITLE
Feat/add platform version fields to devices

### PR DESCRIPTION
**Pull request by t4deu**
Closes the issue #110

### CODE REVIEW

**Comment by juliogreff** [comment]
Try not to close one pull request and open another one for the same feature. You can force-push the branch and the pull request will be updated automatically.

Also, your commit messages *still* don't have any mention of the issue you're trying to fix. Please edit the message of at least one of them (preferrably the first one, so we can squash later if needed) and push again.

**Comment by t4deu** [comment]
Close the old push was not my first option, but to has difficult to remove the old commits, the more simple way was to create a new branch and cherry-pick the higth commits

**Comment by juliogreff** [comment]
```
git reset --hard master
git cherry-pick <commit>
git push -f
```

Do ask for help when you need it.

**Comment by juliogreff** [comment]
Looks good, I'll do some testing today.

Can you squash the commits so they're a nice atomic unit? Also, put some thought into the message.

http://who-t.blogspot.nl/2009/12/on-commit-messages.html

**Comment by t4deu** [comment]
Merged the commits like you asked, and TRIED to create a good commit message, sorry if it's not that good.

**Comment by juliogreff** [comment]
You don't need to change it, but try to keep the first line shorter in the future (72 characters or less) so it's easier to read from the terminal. It doesn't need to say *exactly* what was done, that's what the body of the commit message is for.

I'll do some testing today. Meanwhile, there's some documentation that needs updating. Your new fields are not in the api specs yet.

**Comment by t4deu** [comment]
The api spec is updated now

**Comment by juliogreff** [comment]
We still have two API calls that create/update devices, but their documentation is outdated.

* https://github.com/meumobi/sitebuilder/blob/feat/add-platform-version-fields-to-devices/doc/api_specs.md#post-visitorslogin

* https://github.com/meumobi/sitebuilder/blob/feat/add-platform-version-fields-to-devices/doc/api_specs.md#post-visitorslogin

File: sitebuilder/app/controllers/api/VisitorsController.php [hunk 1]
```diff
@@ -122,20 +106,12 @@ public function update_device()
...
-				'model' => $this->request->get('data:model'),
-				'app_version' => $this->request->get('data:app_version')
-			]);
-			$visitor->addDevice($device);
-		}
+		$deviceData = $this->request->data;
+		$deviceData['uuid'] = $this->request->get('params:uuid');
+
+		$service = new CreateOrUpdateDevice();
+		$service->perform($deviceData, $visitor);
```
Comment (juliogreff) [inline sitebuilder/app/controllers/api/VisitorsController.php]: Since we're at it, this might also benefit from an associative array.

**Comment by juliogreff** [comment]
Tests performed:

* Logging in without a device: **OK**
* Logging in with an invalid device (missing uuid): **NOK** -- warning "Undefined index: uuid"
* Logging in with a new, valid device: **OK**
* Logging in with a different device: **OK**
* Logging in with am updated device: **OK**
* Creating a new device: **OK**
* Creating a device: **OK**

There's also still a [comment that was not addressed](https://github.com/meumobi/sitebuilder/pull/115#discussion_r32433079). After that, good to merge :+1:

**Comment by t4deu** [comment]
I prevent the visitor login if the device is invalid or just log the error?

**Comment by juliogreff** [comment]
I'd say that you can allow the visitor to log in, but definitely return the error as well, so it can be diagnosed instead of silently fail.

**Comment by t4deu** [comment]
I will put the message in the "errors" field of the response, but will be weird if the password of the visitor is expired:
```json
{
    "success": true,
    "token": "385cfc03417e75bd8bc3dafaa92b8f178eb763f0",
    "visitor": {
       ....
    },
    "error": "password expired",
    "errors": "invalid device uuid"
}

**Comment by juliogreff** [comment]
"errors" should be an array, so if the password is also expired, you have them both.

File: config/environments/development.php [hunk 2]
```diff
@@ -11,6 +11,6 @@
 Config::write('Preview.url', 'http://placeholder.int-meumobi.com');
 Config::write('Sites.domain', 'int-meumobi.com');
 
-Config::write('Themes.url', 'http://meu-template-engine.int-meumobi.com/themes.json');
+Config::write('Themes.url', 'http://meu-template-engine.int-meumobilesite.com/themes.json');
```
Comment (juliogreff) [inline config/environments/development.php]: Why are these changes relevant to this PR?

File: sitebuilder/app/controllers/api/VisitorsController.php [hunk 3]
```diff
@@ -35,28 +36,36 @@ protected function loginUser()
...
+			$response = [
+				'success' => true,
+				'errors' => []
+			];
+
 			if ($deviceData) {
-				$device = new VisitorDevice($deviceData);
-				$visitor->addDevice($device);
+				try {
+					$this->createOrUpdateDevice($deviceData, $visitor);
```
Comment (juliogreff) [inline sitebuilder/app/controllers/api/VisitorsController.php]: Why can't we do the same we did with the items? I mean, returning two values, one for the thing that was created, and another for the errors?

File: sitebuilder/app/controllers/api/VisitorsController.php [hunk 4]
```diff
@@ -122,20 +112,11 @@ public function update_device()
...
-				'pushId' => $this->request->get('data:push_id'),
-				'model' => $this->request->get('data:model'),
-				'app_version' => $this->request->get('data:app_version')
-			]);
-			$visitor->addDevice($device);
-		}
+		$deviceData = $this->request->data;
+		$deviceData['uuid'] = $this->request->get('params:uuid');
+
+		$this->createOrUpdateDevice($deviceData, $visitor);
```
Comment (juliogreff) [inline sitebuilder/app/controllers/api/VisitorsController.php]: We're *still* missing the associative array here

File: config/environments/development.php (hunk 2 above)
Comment (t4deu) [inline config/environments/development.php]: No they are not, they are the result of an "git add ." from a developer with little attention :)

File: sitebuilder/app/controllers/api/VisitorsController.php (hunk 4 above)
Comment (t4deu) [inline sitebuilder/app/controllers/api/VisitorsController.php]: I understand that the associative array is good because there is no need to remember the parameters order, but IT not difficult the code readability?
For example if in a long future I need to update the createOrUpdateDevice method, I will need to check every line of it to see what are the available parameters and what they do.

Ps. I'm finally reading "Clean Code", so a lot of doubts are appearing

Comment (t4deu) [inline sitebuilder/app/controllers/api/VisitorsController.php]: The separated method is to remove the code duplication

Comment (juliogreff) [inline sitebuilder/app/controllers/api/VisitorsController.php]: You don't need to check every line. You can assign it to variables in the top of your method. Just like we did in #111.

There's no duplication enough to justify creating a method. Just create the service and call it directly. If you feel that's too much, you should create a static method on the service instead.
